File: src/transfer.ts
// =============================================================================
// CORE-FFX - Forensic File Explorer
// Copyright (c) 2024-2026 CORE-FFX Project Contributors
// Licensed under MIT License - see LICENSE file for details
// =============================================================================

/**
 * File Transfer API - TypeScript bindings for file copy/transfer operations
 * 
 * Provides functions for:
 * - Copying single files or entire directories
 * - Progress tracking with real-time events
 * - Hash verification after copy
 * - Cancellation support for long operations
 */

import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";
import { formatBytes } from "./utils";

// Re-export formatBytes from utils for backwards compatibility
export { formatBytes };

/**
 * Request to start a file transfer operation
 */
export interface TransferRequest {
  /** Source paths to copy */
  sources: string[];
  /** Destination directory */
  destination: string;
  /** Whether to verify files after copying (hash comparison) */
  verify?: boolean;
  /** Hash algorithm for verification ("md5", "sha256", "xxh3", etc.) */
  hashAlgorithm?: string;
  /** Whether to preserve file timestamps */
  preserveTimestamps?: boolean;
  /** Whether to preserve permissions */
  preservePermissions?: boolean;
  /** Whether to overwrite existing files */
  overwrite?: boolean;
  /** Whether to copy directories recursively */
  recursive?: boolean;
  /** Whether to flatten directory structure */
  flatten?: boolean;
  /** Whether sources contain forensic containers (E01, AD1, etc.) that should be treated as logical units */
  containerAware?: boolean;
  /** Number of parallel transfer threads (1-8, default: 4) */
  parallelThreads?: number;
}

/**
 * Preview of files to be transferred
 */
export interface TransferPreview {
  /** Total number of files (may be limited for large directories) */
  total_files: number;
  /** Total number of directories */
  total_directories: number;
  /** Total size in bytes */
  total_bytes: number;
  /** Human-readable size */
  total_size_formatted: string;
  /** List of files with their relative paths and sizes */
  files: TransferFileInfo[];
  /** Whether the preview was limited (more files exist than shown) */
  is_limited?: boolean;
}

/**
 * Information about a file in the transfer preview
 */
export interface TransferFileInfo {
  /** Source path */
  source: string;
  /** Relative path (for destination) */
  relative_path: string;
  /** Size in bytes */
  size: number;
  /** Human-readable size */
  size_formatted: string;
}

/**
 * Phase of a transfer operation
 */
export type TransferPhase = 
  | "Scanning"
  | "Copying"
  | "Verifying"
  | "Completed"
  | "Cancelled"
  | "Failed";

/**
 * Progress information for a transfer operation
 */
export interface TransferProgress {
  /** Unique operation ID */
  operation_id: string;
  /** Current phase of the operation */
  phase: TransferPhase;
  /** Total number of files to transfer */
  total_files: number;
  /** Number of files completed */
  files_completed: number;
  /** Total bytes to transfer */
  total_bytes: number;
  /** Bytes transferred so far */
  bytes_transferred: number;
  /** Current file being processed */
  current_file: string | null;
  /** Current file progress (0.0 - 1.0) */
  current_file_progress: number;
  /** Overall progress percentage (0.0 - 100.0) */
  overall_percent: number;
  /** Transfer rate in bytes per second */
  bytes_per_second: number;
  /** Estimated time remaining in seconds */
  eta_seconds: number | null;
  /** Whether the operation is cancelled */
  cancelled: boolean;
  /** Error message if any */
  error: string | null;
  /** Number of parallel threads being used */
  parallel_threads: number;
  /** Files currently being transferred (for parallel transfers) */
  active_files: string[];
}

/**
 * Result of a single file transfer
 */
export interface FileTransferResult {
  /** Source path */
  source: string;
  /** Destination path */
  destination: string;
  /** File size in bytes */
  size: number;
  /** Whether the transfer succeeded */
  success: boolean;
  /** Error message if failed */
  error: string | null;
  /** Source hash (if computed) */
  source_hash: string | null;
  /** Destination hash (if verified) */
  destination_hash: string | null;
  /** Whether hashes matched (if verified) */
  verified: boolean | null;
}

/**
 * Result of a complete transfer operation
 */
export interface TransferResult {
  /** Unique operation ID */
  operation_id: string;
  /** Whether the overall operation succeeded */
  success: boolean;
  /** Total files processed */
  total_files: number;
  /** Successfully transferred files */
  successful_files: number;
  /** Failed files */
  failed_files: number;
  /** Skipped files (already exist, etc.) */
  skipped_files: number;
  /** Total bytes transferred */
  bytes_transferred: number;
  /** Duration in milliseconds */
  duration_ms: number;
  /** Individual file results */
  files: FileTransferResult[];
  /** Error message if operation failed */
  error: string | null;
}

// =============================================================================
// API Functions
// =============================================================================

/**
 * Preview a transfer operation without executing it
 * 
 * Returns a summary of files that would be transferred.
 * 
 * @param sources - Array of source paths to preview
 * @param recursive - Whether to include subdirectories (default: true)
 * @returns Preview of transfer operation
 * 
 * @example
 * ```typescript
 * const preview = await transferPreview(["/path/to/folder"]);
 * console.log(`Would transfer ${preview.total_files} files (${preview.total_size_formatted})`);
 * ```
 */
export async function transferPreview(
  sources: string[],
  recursive: boolean = true
): Promise<TransferPreview> {
  return invoke<TransferPreview>("transfer_preview", { sources, recursive });
}

/**
 * Start a file transfer operation
 * 
 * Returns the operation ID which can be used to:
 * - Cancel the transfer with `transferCancel()`
 * - Track progress via the "transfer-progress" event
 * 
 * @param request - Transfer configuration
 * @returns Operation ID
 * 
 * @example
 * ```typescript
 * const opId = await transferStart({
 *   sources: ["/path/to/folder"],
 *   destination: "/path/to/destination",
 *   verify: true,
 * });
 * ```
 */
export async function transferStart(request: TransferRequest): Promise<string> {
  return invoke<string>("transfer_start", { request });
}

/**
 * Cancel an active transfer operation
 * 
 * @param operationId - The operation ID returned by `transferStart()`
 * @returns true if the operation was found and cancelled, false otherwise
 */
export async function transferCancel(operationId: string): Promise<boolean> {
  return invoke<boolean>("transfer_cancel", { operationId });
}

/**
 * Get list of active transfer operation IDs
 * 
 * @returns Array of operation IDs
 */
export async function transferListActive(): Promise<string[]> {
  return invoke<string[]>("transfer_list_active");
}

/**
 * Copy a single file with optional verification
 * 
 * Simpler API for single file copies without progress events.
 * 
 * @param source - Source file path
 * @param destination - Destination file path
 * @param verify - Whether to verify the copy with hash comparison
 * @param hashAlgorithm - Hash algorithm to use for verification
 * @returns Transfer result for the single file
 * 
 * @example
 * ```typescript
 * const result = await transferCopyFile(
 *   "/path/to/source.pdf",
 *   "/path/to/destination.pdf",
 *   true // verify
 * );
 * if (result.verified) {
 *   console.log("Copy verified successfully");
 * }
 * ```
 */
export async function transferCopyFile(
  source: string,
  destination: string,
  verify?: boolean,
  hashAlgorithm?: string
): Promise<FileTransferResult> {
  return invoke<FileTransferResult>("transfer_copy_file", {
    source,
    destination,
    verify,
    hash_algorithm: hashAlgorithm,
  });
}

/**
 * Copy an entire directory
 * 
 * @param source - Source directory path
 * @param destination - Destination directory path
 * @param verify - Whether to verify files after copying
 * @param recursive - Whether to copy subdirectories (default: true)
 * @returns Operation ID for tracking progress
 */
export async function transferCopyDirectory(
  source: string,
  destination: string,
  verify?: boolean,
  recursive?: boolean
): Promise<string> {
  return invoke<string>("transfer_copy_directory", {
    source,
    destination,
    verify,
    recursive,
  });
}

/**
 * Calculate the total size of files that would be transferred
 * 
 * @param sources - Array of source paths
 * @param recursive - Whether to include subdirectories
 * @returns Total size in bytes
 */
export async function transferCalculateSize(
  sources: string[],
  recursive: boolean = true
): Promise<number> {
  return invoke<number>("transfer_calculate_size", { sources, recursive });
}

// =============================================================================
// Event Listeners
// =============================================================================

/**
 * Listen for transfer progress events
 * 
 * @param callback - Function called with progress updates
 * @returns Unsubscribe function
 * 
 * @example
 * ```typescript
 * const unlisten = await onTransferProgress((progress) => {
 *   console.log(`${progress.overall_percent.toFixed(1)}% complete`);
 *   console.log(`Current file: ${progress.current_file}`);
 * });
 * 
 * // Later: stop listening
 * unlisten();
 * ```
 */
export async function onTransferProgress(
  callback: (progress: TransferProgress) => void
): Promise<UnlistenFn> {
  return listen<TransferProgress>("transfer-progress", (event) => {
    callback(event.payload);
  });
}

/**
 * Listen for transfer completion events
 * 
 * @param callback - Function called when a transfer completes
 * @returns Unsubscribe function
 * 
 * @example
 * ```typescript
 * const unlisten = await onTransferComplete((result) => {
 *   if (result.success) {
 *     console.log(`Transferred ${result.successful_files} files`);
 *   } else {
 *     console.error(`Transfer failed: ${result.error}`);
 *   }
 * });
 * ```
 */
export async function onTransferComplete(
  callback: (result: TransferResult) => void
): Promise<UnlistenFn> {
  return listen<TransferResult>("transfer-complete", (event) => {
    callback(event.payload);
  });
}

// =============================================================================
// Utility Functions
// =============================================================================

/**
 * Format transfer speed to human-readable string
 * 
 * @param bytesPerSecond - Transfer rate in bytes per second
 * @returns Formatted string (e.g., "125 MB/s")
 */
export function formatSpeed(bytesPerSecond: number): string {
  return `${formatBytes(bytesPerSecond)}/s`;
}

/**
 * Format ETA seconds to human-readable string
 * 
 * @param seconds - Estimated time remaining in seconds
 * @returns Formatted string (e.g., "2m 30s", "1h 15m")
 */
export function formatEta(seconds: number | null | undefined): string {
  if (seconds === null || seconds === undefined || seconds < 0 || !isFinite(seconds)) {
    return "calculating...";
  }
  
  if (seconds < 60) {
    return `${Math.round(seconds)}s`;
  }
  
  const minutes = Math.floor(seconds / 60);
  const remainingSeconds = Math.round(seconds % 60);
  
  if (minutes < 60) {
    return `${minutes}m ${remainingSeconds}s`;
  }
  
  const hours = Math.floor(minutes / 60);
  const remainingMinutes = minutes % 60;
  
  return `${hours}h ${remainingMinutes}m`;
}

File: src-tauri/src/commands/transfer.rs
// =============================================================================
// CORE-FFX - Forensic File Explorer
// Copyright (c) 2024-2026 CORE-FFX Project Contributors
// Licensed under MIT License - see LICENSE file for details
// =============================================================================

//! File Transfer Commands
//!
//! Tauri commands for file copy/transfer operations with progress tracking.

use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::thread;

use once_cell::sync::Lazy;
use serde::{Deserialize, Serialize};
use tauri::{AppHandle, Emitter};
use tracing::{error, info};

use crate::common::transfer::{
    self, execute_transfer, scan_transfer_items_limited, TransferConfig,
    TransferState,
};
use crate::common::hex::format_size_compact;

// =============================================================================
// Global State for Active Transfers
// =============================================================================

/// Active transfer operations that can be cancelled
static ACTIVE_TRANSFERS: Lazy<Mutex<HashMap<String, Arc<TransferState>>>> = 
    Lazy::new(|| Mutex::new(HashMap::new()));

// =============================================================================
// Types
// =============================================================================

/// Request to start a file transfer
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TransferRequest {
    /// Source paths to copy
    pub sources: Vec<String>,
    /// Destination directory
    pub destination: String,
    /// Whether to verify files after copying (hash comparison)
    pub verify: Option<bool>,
    /// Hash algorithm for verification ("md5", "sha256", "xxh3", etc.)
    pub hash_algorithm: Option<String>,
    /// Whether to preserve file timestamps
    pub preserve_timestamps: Option<bool>,
    /// Whether to preserve permissions
    pub preserve_permissions: Option<bool>,
    /// Whether to overwrite existing files
    pub overwrite: Option<bool>,
    /// Whether to copy directories recursively
    pub recursive: Option<bool>,
    /// Whether to flatten directory structure
    pub flatten: Option<bool>,
    /// Whether sources contain forensic containers that should be treated as logical units
    pub container_aware: Option<bool>,
    /// Number of parallel transfer threads (1-8, default: 4)
    pub parallel_threads: Option<u8>,
}

/// Summary of files to be transferred (for preview)
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TransferPreview {
    /// Total number of files (may be limited for large directories)
    pub total_files: u64,
    /// Total number of directories
    pub total_directories: u64,
    /// Total size in bytes
    pub total_bytes: u64,
    /// Human-readable size
    pub total_size_formatted: String,
    /// List of files with their relative paths and sizes
    pub files: Vec<TransferFileInfo>,
    /// Whether the preview was limited (more files exist)
    #[serde(default)]
    pub is_limited: bool,
}

/// Information about a file in the transfer preview
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TransferFileInfo {
    /// Source path
    pub source: String,
    /// Relative path (for destination)
    pub relative_path: String,
    /// Size in bytes
    pub size: u64,
    /// Human-readable size
    pub size_formatted: String,
}

// =============================================================================
// Commands
// =============================================================================

/// Maximum number of files to include in preview (for performance)
const MAX_PREVIEW_FILES: usize = 1000;

/// Preview a transfer operation without executing it
/// 
/// Returns a summary of files that would be transferred.
/// Note: File list is limited to MAX_PREVIEW_FILES for performance.
/// When limited, `is_limited` will be true to indicate more files exist.
#[tauri::command]
pub fn transfer_preview(
    sources: Vec<String>,
    recursive: Option<bool>,
) -> Result<TransferPreview, String> {
    info!("transfer_preview: {} sources, recursive={:?}", sources.len(), recursive);
    
    let recursive = recursive.unwrap_or(true);
    
    // Use limited scan for faster preview - only scan up to MAX_PREVIEW_FILES + 1
    // to detect if there are more files
    let items = scan_transfer_items_limited(&sources, recursive, None, Some(MAX_PREVIEW_FILES + 1))
        .map_err(|e| format!("Failed to scan sources: {}", e))?;
    
    let is_limited = items.iter().filter(|i| !i.is_directory).count() > MAX_PREVIEW_FILES;
    
    let mut total_files: u64 = 0;
    let mut total_directories: u64 = 0;
    let mut total_bytes: u64 = 0;
    let mut files = Vec::with_capacity(MAX_PREVIEW_FILES.min(items.len()));
    
    for item in items {
        if item.is_directory {
            total_directories += 1;
        } else {
            total_files += 1;
            total_bytes += item.size;
            // Only include up to MAX_PREVIEW_FILES in the detailed list
            if files.len() < MAX_PREVIEW_FILES {
                files.push(TransferFileInfo {
                    source: item.source.to_string_lossy().to_string(),
                    relative_path: item.relative_path.to_string_lossy().to_string(),
                    size: item.size,
                    size_formatted: format_size_compact(item.size),
                });
            }
        }
    }
    
    Ok(TransferPreview {
        total_files,
        total_directories,
        total_bytes,
        total_size_formatted: format_size_compact(total_bytes),
        files,
        is_limited,
    })
}

/// Start a file transfer operation
/// 
/// Emits "transfer-progress" events as the transfer progresses.
/// Returns the operation ID which can be used to cancel the transfer.
#[tauri::command]
pub async fn transfer_start(
    app: AppHandle,
    request: TransferRequest,
) -> Result<String, String> {
    info!(
        "transfer_start: {} sources -> {}, container_aware={:?}, parallel_threads={:?}", 
        request.sources.len(), 
        request.destination,
        request.container_aware,
        request.parallel_threads
    );
    
    // Create transfer configuration
    // Clamp parallel_threads to 1-8 range, default to 4
    let threads = request.parallel_threads.unwrap_or(4).clamp(1, 8);
    
    let config = TransferConfig {
        sources: request.sources,
        destination: request.destination,
        verify_after_copy: request.verify.unwrap_or(false),
        hash_algorithm: request.hash_algorithm.or(Some("sha256".to_string())),
        preserve_timestamps: request.preserve_timestamps.unwrap_or(true),
        preserve_permissions: request.preserve_permissions.unwrap_or(true),
        overwrite_existing: request.overwrite.unwrap_or(false),
        recursive: request.recursive.unwrap_or(true),
        flatten: request.flatten.unwrap_or(false),
        container_aware: request.container_aware.unwrap_or(false),
        parallel_threads: threads,
    };
    
    // Create shared state for cancellation
    let state = Arc::new(TransferState::new());
    let operation_id = uuid::Uuid::new_v4().to_string();
    
    // Register the transfer
    {
        let mut transfers = ACTIVE_TRANSFERS.lock()
            .map_err(|e| format!("Failed to acquire transfer lock: {}", e))?;
        transfers.insert(operation_id.clone(), Arc::clone(&state));
    }
    
    let op_id = operation_id.clone();
    let app_handle = app.clone();
    
    // Run transfer in background thread
    thread::spawn(move || {
        let result = execute_transfer(&config, Arc::clone(&state), op_id.clone(), |progress| {
            // Emit progress event
            if let Err(e) = app_handle.emit("transfer-progress", &progress) {
                error!("Failed to emit transfer progress: {}", e);
            }
        });
        
        // Emit completion event
        if let Err(e) = app_handle.emit("transfer-complete", &result) {
            error!("Failed to emit transfer complete: {}", e);
        }
        
        // Remove from active transfers
        if let Ok(mut transfers) = ACTIVE_TRANSFERS.lock() {
            transfers.remove(&op_id);
        }
        
        info!(
            "Transfer {} complete: {} files, {} bytes",
            op_id, result.successful_files, result.bytes_transferred
        );
    });
    
    Ok(operation_id)
}

/// Cancel an active transfer operation
#[tauri::command]
pub fn transfer_cancel(operation_id: String) -> Result<bool, String> {
    info!("transfer_cancel: {}", operation_id);
    
    let transfers = ACTIVE_TRANSFERS.lock()
        .map_err(|e| format!("Failed to acquire transfer lock: {}", e))?;
    if let Some(state) = transfers.get(&operation_id) {
        state.cancel();
        Ok(true)
    } else {
        Ok(false) // Transfer not found or already complete
    }
}

/// Get list of active transfer operations
#[tauri::command]
pub fn transfer_list_active() -> Result<Vec<String>, String> {
    let transfers = ACTIVE_TRANSFERS.lock()
        .map_err(|e| format!("Failed to acquire transfer lock: {}", e))?;
    Ok(transfers.keys().cloned().collect())
}

/// Copy a single file with optional verification
/// 
/// Simpler API for single file copies without progress events.
#[tauri::command]
pub async fn transfer_copy_file(
    source: String,
    destination: String,
    verify: Option<bool>,
    hash_algorithm: Option<String>,
) -> Result<transfer::FileTransferResult, String> {
    info!("transfer_copy_file: {} -> {}", source, destination);
    
    let src_path = std::path::PathBuf::from(&source);
    let dst_path = std::path::PathBuf::from(&destination);
    
    if !src_path.exists() {
        return Err(format!("Source file does not exist: {}", source));
    }
    
    if !src_path.is_file() {
        return Err(format!("Source is not a file: {}", source));
    }
    
    // Create parent directory if needed
    if let Some(parent) = dst_path.parent() {
        std::fs::create_dir_all(parent)
            .map_err(|e| format!("Failed to create destination directory: {}", e))?;
    }
    
    let file_size = src_path.metadata()
        .map_err(|e| format!("Failed to read source metadata: {}", e))?
        .len();
    
    // Copy the file
    let copy_result = transfer::copy_file_with_progress(
        &src_path,
        &dst_path,
        true, // preserve timestamps
        true, // preserve permissions
        None, // no cancellation state
        |_, _| {}, // no progress callback for single file
    );
    
    match copy_result {
        Ok(_) => {
            // Verify if requested - use sha256 as default for forensic integrity
            let algo = hash_algorithm.as_deref().unwrap_or("sha256");
            let (source_hash, dest_hash, verified) = if verify.unwrap_or(false) {
                let src_hash = transfer::compute_file_hash(&src_path, algo).ok();
                let dst_hash = transfer::compute_file_hash(&dst_path, algo).ok();
                let is_verified = match (&src_hash, &dst_hash) {
                    (Some(s), Some(d)) => Some(s == d),
                    _ => None,
                };
                (src_hash, dst_hash, is_verified)
            } else {
                (None, None, None)
            };
            
            if verified == Some(false) {
                return Err("Verification failed - hash mismatch".to_string());
            }
            
            Ok(transfer::FileTransferResult {
                source,
                destination,
                size: file_size,
                success: true,
                error: None,
                source_hash,
                destination_hash: dest_hash,
                verified,
            })
        }
        Err(e) => Err(format!("Copy failed: {}", e)),
    }
}

/// Copy entire directory with optional verification
#[tauri::command]
pub async fn transfer_copy_directory(
    app: AppHandle,
    source: String,
    destination: String,
    verify: Option<bool>,
    recursive: Option<bool>,
) -> Result<String, String> {
    info!("transfer_copy_directory: {} -> {}", source, destination);
    
    let src_path = std::path::PathBuf::from(&source);
    
    if !src_path.exists() {
        return Err(format!("Source does not exist: {}", source));
    }
    
    if !src_path.is_dir() {
        return Err(format!("Source is not a directory: {}", source));
    }
    
    // Use the main transfer API
    let request = TransferRequest {
        sources: vec![source],
        destination,
        verify,
        hash_algorithm: Some("xxh3".to_string()),
        preserve_timestamps: Some(true),
        preserve_permissions: Some(true),
        overwrite: Some(false),
        recursive,
        flatten: Some(false),
        container_aware: Some(false),
        parallel_threads: None, // Use default
    };
    
    transfer_start(app, request).await
}

/// Get the size of files that would be transferred
#[tauri::command]
pub fn transfer_calculate_size(
    sources: Vec<String>,
    recursive: Option<bool>,
) -> Result<u64, String> {
    let recursive = recursive.unwrap_or(true);
    
    transfer::calculate_transfer_size(&sources, recursive)
        .map_err(|e| format!("Failed to calculate size: {}", e))
}
